llvm/lib/CodeGen/CheriBoundAllocas.cpp

```diff
@@ -295,44 +295,60 @@ class CheriBoundAllocas : public ModulePass, public InstVisitor<CheriBoundAlloca
-        Instruction *AllocaI8 =
-          cast<Instruction>(B.CreateBitCast(AI, I8CapTy));
-        SingleBoundedAlloc = B.CreateCall(SetBoundsIntrin, {AllocaI8, Size});
+        Instruction *AllocaI8 = cast<Instruction>(B.CreateBitCast(AI, I8CapTy));
+        Value *SingleBoundedAlloc =
+            B.CreateCall(SetBoundsIntrin, {AllocaI8, Size});
         SingleBoundedAlloc = B.CreateBitCast(SingleBoundedAlloc, AllocaTy);
-      }
-      for (const Use *U : UsesThatNeedBounds) {
-        Instruction *I = cast<Instruction>(U->getUser());
-        if (ReuseSingleIntrinsicCall) {
+        for (const Use *U : UsesThatNeedBounds) {
           const_cast<Use *>(U)->set(SingleBoundedAlloc);
-        } else {
-          if (auto PHI = dyn_cast<PHINode>(I)) {
-            // For PHI nodes we can't insert just before the PHI, instead we
-            // must insert it just before
-            auto BB = PHI->getIncomingBlock(*U);
-            // LLVM_DEBUG(dbgs() << "PHI use coming from"; BB->dump());
-            B.SetInsertPoint(BB->getTerminator());
+        }
+      } else {
+        // Otherwise, we create new intrinsics for every use. This can avoid
+        // stack spills but will result in additional instructions.
+        // When we encounter multiple uses within the same instruction, we need
+        // to ensure that we reuse the same bounded alloca intrinsic. If we
+        // don't do this, we could end up creating invalid PHI nodes where the
+        // PHI node has multiple entries for the same basic block and uses
+        // different incoming values (for an example see
+        // multiple-uses-in-same-instr.ll). This not only avoids invalid IR, but
+        // also avoids unnecessarily creating multiple intrinsic calls e.g.
+        // in cases where a call instruction passes the same IR variable twice.
+        SmallDenseMap<User *, Value *> ReplacedUses;
+        for (const Use *U : UsesThatNeedBounds) {
+          Value *BoundedAlloca;
+          Instruction *I = cast<Instruction>(U->getUser());
+          auto It = ReplacedUses.find(I);
+          if (It == ReplacedUses.end()) {
+            // First use in this instruction -> create a new intrinsic call.
+            if (auto PHI = dyn_cast<PHINode>(I)) {
+              // For PHI nodes we can't insert just before the PHI, instead we
+              // must insert it just before the end of the incoming BB.
+              auto BB = PHI->getIncomingBlock(*U);
+              // LLVM_DEBUG(dbgs() << "PHI use coming from"; BB->dump());
+              B.SetInsertPoint(BB->getTerminator());
+            } else {
+              // Insert just before the use. This should avoid spilling
+              // registers when using an alloca in a different basic block.
+              B.SetInsertPoint(I);
+            }
+            // We need to convert it to an i8* for the intrinisic. Note: we have
+            // to create a new bitcast every time since reusing the same one can
```

Comment:
This seems... odd; bitcasts _should_ be no-ops that get lost way before register allocation...